### PR TITLE
Fix ReadableByteChannel.read usages

### LINKED ISSUE #5088: Helpers.readBytes assumes ReadableByteChannel#read will always read fully
This was caused during exploration of the new s3 feature, https://github.com/deephaven/deephaven-core/pull/4972.

```
caused by java.io.IOException: Expected for bytes, only read 7244 while it expected 38744
        at io.deephaven.parquet.base.util.Helpers.readBytes(Helpers.java:20)
        at io.deephaven.parquet.base.ParquetFileReader.<init>(ParquetFileReader.java:89)
        at io.deephaven.parquet.table.ParquetTools.getParquetFileReaderChecked(ParquetTools.java:1163)
        at io.deephaven.parquet.table.ParquetTools.getParquetFileReader(ParquetTools.java:1132)
        ... 27 more
```

This was easy to reproduce using a small fragment size `io.deephaven.extensions.s3.S3Instructions#fragmentSize`. While not caused by this feature, we should aim to fix this to make sure the new feature works as expected.

### PR DESCRIPTION
**Pull request by devinrsmith**
Fixes #5088

### CODE REVIEW

**Review by malhotrashivam — APPROVED** [review]

File: extensions/arrow/src/main/java/io/deephaven/extensions/arrow/ArrowWrapperTools.java [hunk 1]
```diff
@@ -157,11 +157,12 @@ public static QueryTable readFeather(@NotNull final String path) {
                 channel.position(block.getOffset());
 
                 final ByteBuffer metadataBuf = ByteBuffer.wrap(rawMetadataBuf, 0, block.getMetadataLength());
-                int numRead = channel.read(metadataBuf);
-                if (numRead != block.getMetadataLength()) {
-                    throw new IOException("Unexpected end of input trying to read block " + ii + " of '" + path + "'");
+                while (metadataBuf.hasRemaining()) {
+                    if (channel.read(metadataBuf) == -1) {
```
Comment (nbauernfeind) [inline extensions/arrow/src/main/java/io/deephaven/extensions/arrow/ArrowWrapperTools.java:161]: We should probably throw if `<= 0` to ensure that each call is making progress.

File: extensions/parquet/base/src/main/java/io/deephaven/parquet/base/util/Helpers.java [hunk 2]
```diff
@@ -15,22 +15,26 @@
...
-        }
+        readExact(f, buffer);
         buffer.flip();
         return BytesInput.from(buffer);
     }
 
+    public static void readExact(ReadableByteChannel f, ByteBuffer buffer) throws IOException {
+        final int expected = buffer.remaining();
+        while (buffer.hasRemaining()) {
+            if (f.read(buffer) == -1) {
```
Comment (nbauernfeind) [inline extensions/parquet/base/src/main/java/io/deephaven/parquet/base/util/Helpers.java:31]: Same; probably should complain if we aren't making progress by checking `<= 0`.

**Review by nbauernfeind — APPROVED** [review]
I won't throw my hands in the air if you don't like the suggestion or feel that it would not be correct.

File: extensions/arrow/src/main/java/io/deephaven/extensions/arrow/ArrowWrapperTools.java (hunk 1 above)
Comment (devinrsmith) [reply extensions/arrow/src/main/java/io/deephaven/extensions/arrow/ArrowWrapperTools.java:161]: It's a valid point. Assuming the channel is in blocking mode, we _can_ throw an error on 0 (which _shouldn't_ occur)

> It is guaranteed, however, that if a channel is in blocking mode and there is at least one byte remaining in the buffer then this method will block until at least one byte is read.

I'll make the change.

**Review by nbauernfeind — APPROVED** [review]

**Pull request merged**